data/market.py

```diff
@@ -0,0 +1,176 @@
+
+    @property
+    def invariant(self) -> int:
+        return self._invariant
+
+    @property
+    def base_token_amount(self) -> int:
+        return self._base_token_amount
+
+    @property
+    def quote_token_amount(self) -> int:
+        return self._quote_token_amount
+
+    @property
+    def base_token(self) -> Token:
+        return self._base_token
+
+    @property
+    def quote_token(self) -> Token:
+        return self._quote_token
+
+    @property
+    def invariant(self) -> int:
+        return self._invariant
+
+    def set_invariant(self) -> None:
+        if self._type == "CPF":
+            self._invariant = self._base_token_amount * self._quote_token_amount
+
+    def exchange_rate(self) -> float:
+        return float(self._quote_token_amount / self._base_token_amount)
+
+    def set_exchange_rate(self, value) -> None:
+        pass
+
+    def add_liquidity(self, amount: int) -> None:
+        """Adds liquidity too the pool.
+
+        Args:
+            amount (int): The amount is the value to be added, in units of the quote currency that will be split 50/50 between boths tokens.
+            E.g. if the quote currency is USD, as in a BTC/USD quote, the value is in USD. If the quote currency is something else like BTC/DOT
+            the amount would be the amount of DOTs added to the pool (which will be split 50/50 still).
+        """
+        self._base_token_amount += amount / 2 / self.exchange_rate()
+        self._quote_token_amount += amount / 2
+        self.set_invariant()
+
+    def remove_liquidity(self, amount: int) -> None:
+        """Adds liquidity too the pool.
+
+        Args:
+            amount (int): The amount is the value to be removed, in units of the quote currency that will be split 50/50 between boths tokens.
+            E.g. if the quote currency is USD, as in a BTC/USD quote, the value is in USD. If the quote currency is something else like BTC/DOT
+            the amount would be the amount of DOTs added to the pool (which will be split 50/50 still).
+        """
+        self._base_token_amount -= amount / 2 / self.exchange_rate()
+        self._quote_token_amount -= amount / 2
+        self.set_invariant()
+
+    # TODO: refacture swap and calculate_slippage!
+
+    def exact_output_swap(self, output_token: Token, amount: int) -> float:
+        """Executes an exact output swap.
+
+        Args:
+            amount (int): Amount of 'output_token' to swap (buy from the pool)
+            output_token (Token): The token to receive as the output of the swap.
+
+        Raises:
+            Exception: Raises and error if swap amount is too large.
+
+        Returns:
+            float: _description_
+        """
+        self._base_token_amount, self._quote_token_amount, _ = self.calculate_params(
+            token=output_token, amount= amount, type="exact_output"
+        )
+
+    def exact_input_swap(self, input_token: Token, amount: int) -> float:
```

Comment:
AFAICT this is not yet used anywhere, but the _base_token_amount` and `_quote_token_amount` amounts should be updated in the pool when an actual trade occurs

Reply:
They get updated
```
self._base_token_amount, self._quote_token_amount, _ = self.calculate_params(
            token=output_token, amount=amount, type="exact_output"
        )
       ```